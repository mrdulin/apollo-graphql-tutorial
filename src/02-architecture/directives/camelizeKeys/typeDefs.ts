import { gql } from 'apollo-server';

const typeDefs = gql`
  directive @camelizeKeys on OBJECT
`;

export { typeDefs };
