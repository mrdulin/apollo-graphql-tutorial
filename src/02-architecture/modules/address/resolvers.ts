import { IResolvers } from 'apollo-server';
import { IAppContext } from '../../server';

const resolvers: IResolvers = {
  Query: {
    addressById: (_, { id }, { dataSources }: IAppContext) => {
      return dataSources.address.findById(id);
    },
  },
};

export { resolvers };
