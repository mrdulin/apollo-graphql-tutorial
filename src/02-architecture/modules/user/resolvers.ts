import { IResolvers } from 'apollo-server';
import { IAppContext } from '../../server';
import fetch from 'node-fetch';
import { logger } from '../../../util';

const resolvers: IResolvers = {
  Query: {
    userById: async (_, { id }, { dataSources }: IAppContext) => {
      // test GCP cloud trace
      const response = await fetch('https://api.itbook.store/1.0/search/mongodb').then((res) => res.json());
      logger.debug('HTTP request for testing GCP cloud trace', { arguments: { response } });
      return dataSources.user.findById(id);
    },
    users: (_, __, { dataSources }: IAppContext) => {
      return dataSources.user.findAll();
    },
  },
  User: {
    userAddress: (source, _, { dataSources }: IAppContext) => {
      return dataSources.address.findById(source.userAddressId);
    },
  },
};

export { resolvers };
