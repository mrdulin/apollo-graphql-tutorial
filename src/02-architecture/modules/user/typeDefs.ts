import { gql } from 'apollo-server';

const typeDefs = gql`
  type User @camelizeKeys {
    userId: ID!
    userNme: String
    userEmail: String!
    userAddress: Address
  }

  extend type Query {
    userById(id: ID!): User
    users: [User]!
  }
`;

export { typeDefs };
