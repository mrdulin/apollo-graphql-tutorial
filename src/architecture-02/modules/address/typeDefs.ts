import { gql } from 'apollo-server';

const typeDefs = gql`
  type Address {
    addressId: ID!
    addressCountry: String
    addressState: String
    addressCity: String
  }

  extend type Query {
    addressById(id: ID!): Address
  }
`;

export { typeDefs };
