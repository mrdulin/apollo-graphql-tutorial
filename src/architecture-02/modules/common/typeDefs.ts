import { gql } from 'apollo-server';

const typeDefs = gql`
  type CommonResponse {
    code: Int!
    message: String!
  }
`;

export { typeDefs };
