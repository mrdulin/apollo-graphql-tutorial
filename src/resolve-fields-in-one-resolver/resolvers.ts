import { IResolvers } from 'graphql-tools';

const resolvers: IResolvers = {
  Query: {
    async postById(_, { id }, { memoryDB }) {
      const postFound = memoryDB.posts.find((post) => post.postId === id);
      if (postFound) {
        const postAuthor = memoryDB.users.find((user) => user.userId === postFound.postAuthorId);
        postFound.postAuthor = postAuthor;
      }
      return postFound;
    },
    async posts(_, __, { memoryDB }) {
      return memoryDB.posts.map((post) => {
        post.postAuthor = memoryDB.users.find((user) => user.userId === post.postAuthorId);
        return post;
      });
    },
  },
};

export { resolvers };
