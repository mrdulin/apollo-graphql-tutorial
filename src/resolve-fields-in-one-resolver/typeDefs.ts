import { gql } from 'apollo-server';

const typeDefs = gql`
  type User {
    userId: ID!
    userNme: String
    userEmail: String
  }

  type Post {
    postId: ID!
    postTitle: String
    postCreatedAt: String
    postAuthor: User
  }

  type Query {
    postById(id: ID!): Post
    posts: [Post]!
  }
`;

export { typeDefs };
