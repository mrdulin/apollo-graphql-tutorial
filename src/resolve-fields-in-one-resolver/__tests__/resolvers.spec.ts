import { createTestClient } from 'apollo-server-testing';
import { resolvers } from '../resolvers';
import { typeDefs } from '../typeDefs';
import { ApolloServer, gql } from 'apollo-server';

const mockedMemoryDB = {
  posts: {
    find: jest.fn(),
    map: jest.fn(),
  },
  users: {
    find: jest.fn(),
  },
};

const server = new ApolloServer({
  typeDefs,
  resolvers,
  context: { memoryDB: mockedMemoryDB },
});
const testClient = createTestClient(server);

const Fragment = {
  PostWithAuthor: `
    fragment PostWithAuthor on Post{
      postId
      postTitle
      postCreatedAt
      postAuthor {
        userId
        userNme
        userEmail
      }
    }
  `,
};

const Query = {
  postById: gql`
    query PostByIdWithAuthor($id: ID!) {
      postById(id: $id) {
        ...PostWithAuthor
      }
    }

    query PostByIdWithoutAuthor($id: ID!) {
      postById(id: $id) {
        postId
        postTitle
        postCreatedAt
      }
    }

    ${Fragment.PostWithAuthor}
  `,
  posts: gql`
    query Posts {
      posts {
        ...PostWithAuthor
      }
    }

    ${Fragment.PostWithAuthor}
  `,
};

const mockedPost = {
  postId: '222',
  postTitle: 'apollo graphql',
  postCreateAt: 'Fri, 06 Sep 2019 10:56:44 GMT',
  postAuthorId: '333',
};

const mockedUser = {
  userId: '333',
  userNme: 'mrdulin',
  userEmail: 'mrdulin@example.com',
};

const mockedPosts = [
  {
    ...mockedPost,
    postAuthor: mockedUser,
  },
];

describe('resolve-fields-in-one-resolver', () => {
  describe('resolvers', () => {
    describe('#postById', () => {
      beforeEach(() => {
        mockedMemoryDB.posts.find.mockReset();
        mockedMemoryDB.users.find.mockReset();
      });
      it('should query post by id with author correctly', async () => {
        mockedMemoryDB.posts.find.mockReturnValueOnce(mockedPost);
        mockedMemoryDB.users.find.mockReturnValueOnce(mockedUser);
        const actualValue = await testClient.query({
          query: Query.postById,
          variables: { id: 1 },
          operationName: 'PostByIdWithAuthor',
        });
        expect(actualValue).toMatchSnapshot();
        expect(mockedMemoryDB.posts.find).toBeCalledTimes(1);
        expect(mockedMemoryDB.users.find).toBeCalledTimes(1);
      });

      it('should query post by id without author correctly, but it will perform a user query from db', async () => {
        mockedMemoryDB.posts.find.mockReturnValueOnce(mockedPost);
        mockedMemoryDB.users.find.mockReturnValueOnce(mockedUser);
        const actualValue = await testClient.query({
          query: Query.postById,
          variables: { id: 1 },
          operationName: 'PostByIdWithoutAuthor',
        });
        expect(actualValue).toMatchSnapshot();
        expect(mockedMemoryDB.posts.find).toBeCalledTimes(1);
        expect(mockedMemoryDB.users.find).toBeCalledTimes(1);
      });
    });

    describe('#posts', () => {
      beforeEach(() => {
        mockedMemoryDB.posts.map.mockReset();
      });

      it('should query posts correctly', async () => {
        mockedMemoryDB.posts.map.mockReturnValueOnce(mockedPosts);
        const actualValue = await testClient.query({ query: Query.posts, operationName: 'Posts' });
        expect(actualValue).toMatchSnapshot();
        expect(mockedMemoryDB.posts.map).toBeCalledTimes(1);
      });
    });
  });
});
