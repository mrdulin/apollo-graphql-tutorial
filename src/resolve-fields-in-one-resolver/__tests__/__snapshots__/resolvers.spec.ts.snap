// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolve-fields-in-one-resolver resolvers #postById should query post by id with author correctly 1`] = `
Object {
  "data": Object {
    "postById": Object {
      "postAuthor": Object {
        "userEmail": "mrdulin@example.com",
        "userId": "333",
        "userNme": "mrdulin",
      },
      "postCreatedAt": null,
      "postId": "222",
      "postTitle": "apollo graphql",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`resolve-fields-in-one-resolver resolvers #postById should query post by id without author correctly, but it will perform a user query from db 1`] = `
Object {
  "data": Object {
    "postById": Object {
      "postCreatedAt": null,
      "postId": "222",
      "postTitle": "apollo graphql",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`resolve-fields-in-one-resolver resolvers #posts should query posts correctly 1`] = `
Object {
  "data": Object {
    "posts": Array [
      Object {
        "postAuthor": Object {
          "userEmail": "mrdulin@example.com",
          "userId": "333",
          "userNme": "mrdulin",
        },
        "postCreatedAt": null,
        "postId": "222",
        "postTitle": "apollo graphql",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
